src/core/shutdown.c

```diff
@@ -159,6 +164,97 @@ static int switch_root_initramfs(void) {
         return switch_root("/run/initramfs", "/oldroot", false, MS_BIND);
 }
 
+/* Read the following fields from /proc/meminfo:
+ *
+ *  NFS_Unstable
+ *  Writeback
+ *  Dirty
+ *
+ * Return true if the sum of these fields is greater than the previous
+ * value input. For all other issues, report the failure and indicate that
+ * the sync is not making progress.
+ */
+static bool sync_making_progress(unsigned long long *prev_dirty) {
+        _cleanup_fclose_ FILE *f = NULL;
+        char line[LINE_MAX];
+        bool r = false;
+        unsigned long long val = 0;
+
+        f = fopen("/proc/meminfo", "re");
+        if (!f)
+                return log_warning_errno(errno, "Failed to open /proc/meminfo: %m");
+
+        FOREACH_LINE(line, f, log_warning_errno(errno, "Failed to parse /proc/meminfo: %m")) {
+                unsigned long long ull = 0;
+
+                if (!first_word(line, "NFS_Unstable:") && !first_word(line, "Writeback:") && !first_word(line, "Dirty:"))
+                        continue;
+                if (sscanf(line, "%*s %llu %*s", &ull) != 1) {
+                        log_warning_errno(errno, "Failed to parse /proc/meminfo: %m");
```

Comment:
as mentioned, sscanf() won't set errno if it reads fewer arguments than requested. It will set this only on IO errors. Hence we shouldn't reference it unconditionally here. Maybe this:

```c
errno = 0
if (sscanf(line, …) != 1) {
        if (errno != 0)
                log_warning_errno(errno, "Failed to parse…:%m");
        else
                log_warning("Failed to parse, …");
}
```